R/RtGam.R

```diff
@@ -26,6 +93,95 @@ RtGam <- function(cases,
+#' When increasing `k` does not improve the default model diagnostics, manual
+#' model checking via [mgcv::gam.check()] may be needed. Also see
+#' [mgcv::choose.k] for guidance.
+#'
+#' ## Slow model fits
+#' [RtGam] models usually fit faster when the model has less flexibility (lower
+#' values of `k`). The guess from [dimensionality_heuristic()] leans toward
+#' providing excess degrees of freedom, so model fits may take a little longer
+#' than needed. If models are taking a long time to converge, it would be
+#' reasonable to set `k` to a small value, checking for convergence, and
+#' increasing `k` if needed until the model convergences. This approach may or
+#' may not be faster than simply waiting for a model with a higher `k` to fit.
+#'
+#' ## Very wiggly data
+#' If running models in a setting where the data seem quite wiggly, exhibiting
+#' sharp jumps or drops, a model with more flexibility than normal may be
+#' needed. `k` should be increased to the maximum possible value. When running
+#' pre-set models in production, it would also be reasonable to fix the value
+#' of `k` above the default. Because GAMs penalize model wiggliness, the fit to
+#' both wiggly and non-wiggly data is likely to be satisfactory, at the cost of
+#' increased runtime.
+#'
+#' # Implementation details
+#' The algorithm to pick `k` is a piecewise function. When \eqn{n \le 10}, then
+#' the chosen value is \eqn{n}. When \eqn{n > 10}, then the selected value is
+#' \eqn{ \lceil \sqrt{10n} \rceil }.
+#' This approach is loosely inspired by Ward et al., 2021. As in Ward et al.,
+#' the degrees of freedom of the spline is set to a reasonably high value to
+#' avoid oversmoothing. The basis dimension increases with the length of the
+#' timeseries. The scaled square root of the dimension of the data is used to
+#' allow for the higher setup cost of the base model while still increasing the
+#' available degrees of freedom when the length of the timeseries increases.
+#'
+#' @param n An integer, the dimension of the data.
+#' @return An integer, the proposed _total_ basis dimensionality available to
+#'   the [RtGam] model.
+#' @references Ward, Thomas, et al. "Growth, reproduction numbers and factors
+#'   affecting the spread of SARS-CoV-2 novel variants of concern in the UK from
+#'   October 2020 to July 2021: a modelling analysis." BMJ open 11.11 (2021):
+#'   e056636.
+#' @seealso [RtGam()] for the use-case and additional documentation as well as
+#'  [mgcv::choose.k] for more general guidance from `mgcv`.
+#' @export
+#' @examples
+#' cases <- 1:10
+#' k <- dimensionality_heuristic(length(cases))
+dimensionality_heuristic <- function(n) {
+  # Input checks
+  rlang::check_required(n, "n", call = rlang::caller_env())
+  check_vector(n)
+  check_integer(n)
+  check_no_missingness(n)
+  check_elements_above_min(n, "n", min = 1)
+  check_vector_length(length(n), "n", min = 1, max = 1)
+
+  if (n < 10) {
```

Comment:
I'm not sure I'm following? The user is able to bypass this function to set `k` directly in `RtGam()`. Are you suggesting some additional argument to this helper function making the default guess change in some way?

Reply:
Something like the function being `sqrt(10 * n))` and the user could set some aspect of that? e.g. `sqrt(x * n)` where `x` is an input.

If so, then we are providing less of a steer for the user. Of course it could be with a default `x = 10`.

Reply:
I'm not sure I love the idea of adding yet another knob to twist here. My philosophy on these helper functions is that they are meant to provide a steer. A sophisticated user can (and should!) modify `k` in the call to `RtGam()`.

Instead, I'd like to provide documentation on when and how `k` should be modified from the default and support the user in making that change directly. I've attempted to modify things in that direction with these revisions.

Please re-open if you have a problem with any of the above!